### PR TITLE
Clear other messages before dry-run to get only the ones produced during

### PR DESCRIPTION
**Pull request by franciscoaguirre**
The dry-run shows in `forwarded_xcms` all the messages in the queues at the time of calling the API.
Each time the API is called, the result could be different.
You could get messages even if you dry-run something that doesn't send a message, like a `System::remark`.

This PR fixes this by clearing the message queues before doing the dry-run, so the only messages left are the ones the users of the API actually care about.

### CODE REVIEW

File: polkadot/xcm/xcm-builder/src/routing.rs [hunk 1]
```diff
@@ -149,12 +153,20 @@ impl EnsureDelivery for Tuple {
 /// Inspects messages in queues.
 /// Meant to be used in runtime APIs, not in runtimes.
 pub trait InspectMessageQueues {
+	/// Clear the queues, so `Self::get_messages()` can return only the desired events.
```
Comment (acatangiu) [inline polkadot/xcm/xcm-builder/src/routing.rs:156]:
```suggestion
	/// Clear the queues at the beginning of Runtime API call, so that subsequent
	/// `Self::get_messages()` will return only messages generated by said Runtime API call.
```

File: polkadot/xcm/pallet-xcm/src/lib.rs [hunk 2]
```diff
@@ -2457,10 +2457,12 @@ impl<T: Config> Pallet<T> {
 		<RuntimeCall as Dispatchable>::RuntimeOrigin: From<OriginCaller>,
 	{
 		crate::Pallet::<Runtime>::set_record_xcm(true);
+		Router::clear_messages(); // Clear other messages in queues.
 		frame_system::Pallet::<Runtime>::reset_events(); // To make sure we only record events from current call.
```
Comment (acatangiu) [inline polkadot/xcm/pallet-xcm/src/lib.rs:2461]: style consistency

```suggestion
		// Clear other messages in queues,
		Router::clear_messages();
		// and reset events to make sure we only record effects from current call.
		frame_system::Pallet::<Runtime>::reset_events();
```

File: polkadot/runtime/common/src/xcm_sender.rs [hunk 3]
```diff
@@ -141,6 +141,12 @@ where
 }
 
 impl<T: dmp::Config, W, P> InspectMessageQueues for ChildParachainRouter<T, W, P> {
+	fn clear_messages() {
+		let prefix = dmp::DownwardMessageQueues::<T>::final_prefix();
+		// Best effort.
+		let _ = frame_support::storage::unhashed::clear_prefix(&prefix, None, None);
```
Comment (acatangiu) [inline polkadot/runtime/common/src/xcm_sender.rs:147]: Why is this necessary? looks hacky :)

Comment (franciscoaguirre) [reply polkadot/runtime/common/src/xcm_sender.rs:147]: Yes, there's a function [clear](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html#method.clear) on StorageMap that is the nicer API for deleting but it's meant to be used in the runtime, so you **have to** pass in a limit and then iterate until you're done. There's another function [remove_all](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html#method.remove_all) which can remove everything without a limit but it's marked as deprecated. What I ended up doing is using the implementation from `remove_all`.

What I could do is put this under a new method `remove_all_no_consensus` under StorageMap that's meant to not be used in the runtime.

File: polkadot/xcm/xcm-builder/src/routing.rs [hunk 4]
```diff
@@ -149,12 +153,21 @@ impl EnsureDelivery for Tuple {
 /// Inspects messages in queues.
 /// Meant to be used in runtime APIs, not in runtimes.
 pub trait InspectMessageQueues {
+	/// Clear the queues at the beginning of Runtime API call, so that subsequent
+	/// `Self::get_messages()` will return only messages generated by said Runtime API.
+	fn clear_messages() {}
```
Comment (franciscoaguirre) [inline polkadot/xcm/xcm-builder/src/routing.rs:158]: Had to add a default implementation so it's not a breaking change since this is a public trait.

Comment (franciscoaguirre) [reply polkadot/xcm/xcm-builder/src/routing.rs:158]: So we can backport it to stable2409

File: polkadot/runtime/common/src/xcm_sender.rs (hunk 3 above)
Comment (ggwpez) [reply polkadot/runtime/common/src/xcm_sender.rs:147]: You can also put `u32::MAX` as limit?

File: polkadot/xcm/xcm-builder/src/routing.rs (hunk 4 above)
Comment (ggwpez) [reply polkadot/xcm/xcm-builder/src/routing.rs:158]: You could add the default impl specifically only in the Backport MR. Then it would not have a default on master.

Comment (franciscoaguirre) [reply polkadot/xcm/xcm-builder/src/routing.rs:158]: Sounds good

**Review by ggwpez — APPROVED** [review]

File: prdoc/pr_5581.prdoc [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+title: Clear other messages before dry-running
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      The DryRunApi.dry_run_call and DryRunApi.dry_run_xcm functions just returned
+      in forwarded_xcms all the existing messages in the queues at the time.
+      Now messages are cleared before dry-running, meaning only the messages produced
+      by the call (or xcm) will be returned.
```
Comment (acatangiu) [inline prdoc/pr_5581.prdoc:12]:
```suggestion
      The DryRunApi.dry_run_call and DryRunApi.dry_run_xcm functions used to populate
      `forwarded_xcms` with all the existing messages in the queues at the time.
      Now, existing (irrelevant) messages are cleared when dry-running, meaning only the
      messages produced by the dry-run call (or xcm) will be returned in `forwarded_xcms`.
```

**Review by acatangiu — APPROVED** [review]
